File: recursion/02_Josephus/02_Josephus.js
const testCases = [
  { n: 1, k: 1 },
  { n: 2, k: 1 },
  { n: 2, k: 2 },
  { n: 3, k: 1 },
  { n: 3, k: 2 },
  { n: 3, k: 3 },
  { n: 3, k: 4 },
  { n: 4, k: 5 },
  { n: 4, k: 8 },
];

(function () {
  testCases.forEach((testcase, i) => {
    console.log(`n= ${testcase.n}, k= ${testcase.k}`);
    console.log(`testcase ${i}:  ${findTheWinner(testcase.n, testcase.k)}`);
    console.log(`testcase ${i}:  ${findTheWinner2(testcase.n, testcase.k)}`);
    console.log(`testcase ${i}:  ${findTheWinner3(testcase.n, testcase.k)}`);
  });
})();

// Approach 0 : T O(n^2) | S O(n)
function findTheWinner(n, k) {
  let arr = Array.from({ length: n }, (_, i) => i + 1);

  function helper(arr, startIndex) {
    // Base case
    if (arr.length === 1) return arr[0];

    // Recursion case
    let indexToRemove = (startIndex + k - 1) % arr.length;
    arr.splice(indexToRemove, 1);
    helper(arr, indexToRemove);
  }

  return helper(arr, 0);
}

// Approach 1: T O(n) | S O(n)
function findTheWinner2(n, k) {
  function josephus(n) {
    // Base case
    if (n === 1) return 0;

    // Recursive case
    return josephus(n - 1 + k) % n;
  }
  return josephus(n) + 1;
}

// Approach 2: T O(n) | S O(1) / iterative
function findTheWinner3(n, k) {
  let survivor = 0;

  for (let i = 2; i <= n; i++) {
    survivor = (survivor + k) % i;
  }
  return survivor + 1;
}

File: recursion/02_Josephus/go/02_Josephus.go
package main

import "fmt"

func main() {
	testCases := []struct {
		n int
		k int
	}{
		{1, 1},
		{2, 1},
		{2, 2},
		{3, 1},
		{3, 2},
		{3, 3},
		{3, 4},
		{4, 5},
		{4, 8},
	}

	for i, testcase := range testCases {
		fmt.Printf("TestCase %v: \n", i)
		fmt.Printf("findTheWinner1: %v\n", findTheWinner0(testcase.n, testcase.k))
		fmt.Printf("findTheWinner2: %v\n", findTheWinner1(testcase.n, testcase.k))
		fmt.Printf("findTheWinner3: %v\n", findTheWinnerRecursive(testcase.n, testcase.k))
		fmt.Printf("findTheWinner4: %v\n", findTheWinner2(testcase.n, testcase.k))
	}
}

// Approach 0: T O(n^2) | S O(n)
func findTheWinner0(n int, k int) int {

	// Creating array from 1 to n
	arr := make([]int, n)
	for i := 0; i < n; i++ {
		arr[i] = i + 1
	}

	// Define the recursion helper function
	var helper func([]int, int) int
	helper = func(arr []int, startIndex int) int {
		// Base case
		if len(arr) == 1 {
			return arr[0]
		}

		// Recursion case
		indexToRemove := (startIndex + k - 1) % len(arr)
		arr = append(arr[:indexToRemove], arr[indexToRemove+1:]...)
		return helper(arr, indexToRemove)
	}

	return helper(arr, 0)
}

// Approach 1: T O(n) | S O(n)
func findTheWinner1(n int, k int) int {

	var josephus func(int) int
	josephus = func(n int) int {
		//Base case
		if n == 1 {
			return 0
		}
		// Recursive case
		return (josephus(n-1) + k) % n

	}

	return josephus(n) + 1
}

// Approach 1: T O(n) | S O(n) / cleaner
func findTheWinnerRecursive(n int, k int) int {
	if n == 1 {
		return 1
	}
	return (findTheWinnerRecursive(n-1, k)+k-1)%n + 1
}

// Approach 2: T O(n) | S O(1) / iterative
func findTheWinner2(n int, k int) int {

	var survivor int

	for i := 2; i <= n; i++ {
		survivor = (survivor + k) % i
	}

	return survivor + 1
}
